src/protocol/oprf/shuffle/mod.rs

```diff
@@ -0,0 +1,403 @@
+        batch_size,
+    )
+    .await?;
+
+    let mut y_2: Vec<S> = add_single_shares(y_1, z_31).collect();
+
+    let ctx_perm = ctx.narrow(&OPRFShuffleStep::ApplyPermutations);
+    let (mut rng_perm_l, mut rng_perm_r) = ctx_perm.prss_rng();
+    apply_permutation(&mut rng_perm_r, &mut y_2);
+
+    let mut y_3: Vec<S> = add_single_shares_in_place(y_2, z_23);
+    apply_permutation(&mut rng_perm_l, &mut y_3);
+
+    let c_hat_2: Vec<S> = add_single_shares(y_3.iter(), a_hat.iter()).collect();
+    let ((), c_hat_1): ((), Vec<S>) = future::try_join(
+        send_to_peer(
+            ctx,
+            &OPRFShuffleStep::TransferCHat,
+            Direction::Left,
+            c_hat_2.clone(),
+        ),
+        receive_from_peer(
+            ctx,
+            &OPRFShuffleStep::TransferCHat,
+            Direction::Left,
+            batch_size,
+        ),
+    )
+    .await?;
+
+    let c_hat = add_single_shares(c_hat_1, c_hat_2);
+    let res = combine_single_shares(c_hat, a_hat).collect::<Vec<_>>();
+    Ok(res)
+}
+
+fn add_single_shares<A, B, S, L, R>(l: L, r: R) -> impl Iterator<Item = S>
+where
+    A: Add<B, Output = S>,
+    L: IntoIterator<Item = A>,
+    R: IntoIterator<Item = B>,
+{
+    l.into_iter().zip(r).map(|(a, b)| a + b)
+}
+
+fn add_single_shares_in_place<S, R>(mut items: Vec<S>, r: R) -> Vec<S>
+where
+    S: AddAssign<S>,
+    R: IntoIterator<Item = S>,
+{
+    items
+        .iter_mut()
+        .zip(r)
+        .for_each(|(item, rhs)| item.add_assign(rhs));
+    items
+}
+
+// --------------------------------------------------------------------------- //
+
+fn combine_single_shares<S, Il, Ir>(l: Il, r: Ir) -> impl Iterator<Item = AdditiveShare<S>>
+where
+    S: SharedValue,
+    Il: IntoIterator<Item = S>,
+    Ir: IntoIterator<Item = S>,
+{
+    l.into_iter()
+        .zip(r)
+        .map(|(li, ri)| AdditiveShare::new(li, ri))
+}
+
+fn generate_random_tables_with_peers<'a, C, S>(
+    batch_size: usize,
+    narrow_ctx: &'a C,
+) -> (impl Iterator<Item = S> + 'a, impl Iterator<Item = S> + 'a)
+where
+    C: Context,
+    Standard: Distribution<S>,
+    S: 'a,
+{
+    let (rng_l, rng_r) = narrow_ctx.prss_rng();
+    let with_left = rng_l.sample_iter(Standard).take(batch_size);
+    let with_right = rng_r.sample_iter(Standard).take(batch_size);
+    (with_left, with_right)
+}
+
+fn generate_random_table_solo<'a, C, S>(
+    batch_size: usize,
+    narrow_ctx: &'a C,
+    peer: Direction,
+) -> impl Iterator<Item = S> + 'a
+where
+    C: Context,
+    Standard: Distribution<S>,
+    S: 'a,
+{
+    let rngs = narrow_ctx.prss_rng();
+    let rng = match peer {
+        Direction::Left => rngs.0,
+        Direction::Right => rngs.1,
+    };
+
+    rng.sample_iter(Standard).take(batch_size)
+}
+
+// ---------------------------- helper communication ------------------------------------ //
+
+async fn send_to_peer<C, S>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    items: Vec<S>,
```

Comment:
this contract causes extra clones of the size `N` on the caller's side. It is possible to avoid by taking an iterator of `S: SharedValue`.